File: Assets/Scripts/Game/Gameplay/Actors/RespawnSystem.js
var SpawnPoint : Vector3;
var Player : Transform;
var MainCamera : Camera;
var FadeWaitTime : int;
var FootDist : float;  //distance below player to check for ground
var WaterUpgrade : boolean ;
var WaterTime : float; //frames you're allowed to spend on water
var WaterTimeMax : float; //frames you're allowed to spend on water
var Respawning : boolean;

function Start(){
	SpawnPoint = Player.position;
	WaterTime = 0;
	Respawning = false;
}

function Update(){
	if (!Respawning){
	var hit : RaycastHit;
	var touchingWater : boolean;
	touchingWater = false;
		if (Physics.Raycast (Player.position, -Vector3.up, hit)) {
		if (hit.distance<=FootDist){
			if ((hit.transform.tag!="Grind") && (hit.transform.tag!="Water") && (hit.transform.tag!="NoRespawn") && (hit.transform.tag!="Respawn")){
				SpawnPoint = Player.position;
				//Debug.Log(hit.transform.tag);
			}
			if ((hit.transform.tag=="Water") && (!WaterUpgrade)){
				touchingWater = true;
			}
			}
		}
	if (touchingWater){
		if (WaterTime<WaterTimeMax){WaterTime += 1;}
		else{WaterTime = 0; Respawn();}
	}
	else{
		if (WaterTime > 0){WaterTime -= 1;}
	}
	}
}

function OnTriggerEnter (other : Collider) {
	Debug.Log(other.transform);
		if(other.transform.tag == "Respawn"){
			Respawn();
		}
	}

function Respawn()
	{
		//Player.rigidbody.isKinematic = true;
		Debug.Log("respawning...");
		Respawning = true;
		MainCamera.SendMessage("fadeOut");
		yield WaitForSeconds(FadeWaitTime);
		MainCamera.SendMessage("fadeIn");
		Player.position = SpawnPoint;
		Respawning = false;
		//Player.rigidbody.isKinematic = false;
	}

File: Assets/Scripts/RespawnSystem.cs
using UnityEngine;
using System.Collections;

public class RespawnSystem : MonoBehaviour {
	public Vector3 spawnPoint;
	public Transform player;
	public Camera mainCamera;
	public int fadeWaitTime;
	public float footDist;		//distance below player to check for ground
	public bool waterUpgrade;
	public float waterTime; 	//frames you're allowed to spend on water
	public float waterTimeMax; 	//frames you're allowed to spend on water
	public bool isRespawning;
	
	private MonoBehaviour footC;
	private MonoBehaviour boardC;
	private MonoBehaviour ragdollC;
	
	// Use this for initialization
	void Start () {
		spawnPoint = player.position;
		waterTime = 0;
		isRespawning = false;
		
		footC = gameObject.GetComponent<FootController>();
		boardC = gameObject.GetComponent<BoardController>();
		ragdollC = gameObject.GetComponent<RagdollController>();
	}
	
	// Update is called once per frame
	void Update () {
		if (!isRespawning){
			RaycastHit hit;
			bool touchingWater = false;
			Debug.DrawRay(transform.position, -Vector3.up, Color.magenta);
			if (Physics.Raycast (player.position, -Vector3.up, out hit, footDist/*, ~100000000*/)){				//if (hit.distance<=footDist){
				//print ("hit distance:" +hit.distance);
				/*bool isNoke = false;
				Transform par = hit.transform;
				while(par!=null){
						if (par.gameObject.tag == "Player"){isNoke = true; break;}
						par = par.parent;
					}
					if ((!isNoke) && (hit.transform.tag != "Grind") && (hit.transform.tag != "Water") && 
						(hit.transform.tag != "NoRespawn") && (hit.transform.tag != "Respawn"))	{*/
					if(hit.transform.tag == "Terrain"){
						spawnPoint = player.position;
					}
					
					if ((hit.transform.tag == "Water") && (!waterUpgrade)){
						touchingWater = true;
					}
				//}
			}
			if (touchingWater && footC.enabled == true){
				if (waterTime < waterTimeMax)
					waterTime += 1;
				else {
					waterTime = 0;
					Respawn();
				}
			}
			else if(waterTime > 0){
				waterTime -= 1;
			}
		}
	}
	
	IEnumerator Wait(){
		yield return new WaitForSeconds(fadeWaitTime);
	}
	
	void OnTriggerEnter (Collider other){
		//Debug.Log("hit respawner");
		//GetComponent <Animator>().renderer.enabled = false; //Is .renderer needed?
		if(other.transform.tag == "Respawn"){
			Respawn();
		}
	}
	
	public void Respawn(){
		//Debug.log ("respawning");
		isRespawning = true;
		mainCamera.SendMessage("flash");
		//StartCoroutine(Wait());
		//mainCamera.SendMessage("fadeIn");
		player.position = spawnPoint;
		isRespawning = false;
	}
}
